File: eldoria/world/player_character_factory.py
"""Builds a fresh level-1 PlayerCharacter.

Rolls base ability scores (3d6, race-modified), derives starting combat
stats with the same formulas level_progression uses at level 1, and starts
every base (non trainer-locked) skill at a level shaped by class
specialization and race affinity. Trainer-locked skills start unknown except
for a class's one free signature skill, if it has one.
"""
from __future__ import annotations

import random

from eldoria.models import CharacterClass, DiceFormula, DieType, PlayerCharacter, Race, Skill, SkillType, StatBlock
from eldoria.world import stat_generator as sg
from eldoria.world.skill_progression import MAX_SKILL_LEVEL

_BASE_SKILL_FLOOR = 15
_CLASS_PRIMARY_BONUS = 20
_SIGNATURE_SKILL_START = 20


def create(name: str, race: Race, character_class: CharacterClass, rng: random.Random) -> PlayerCharacter:
    strength = max(1, DiceFormula(3, DieType.D6, 0).roll(rng) + race.ability_modifiers.strength)
    agility = max(1, DiceFormula(3, DieType.D6, 0).roll(rng) + race.ability_modifiers.agility)
    willpower = max(1, DiceFormula(3, DieType.D6, 0).roll(rng) + race.ability_modifiers.willpower)
    str_mod = StatBlock.modifier_of(strength)
    agi_mod = StatBlock.modifier_of(agility)

    max_health = max(10, DiceFormula(2, DieType.D8, str_mod * 2).roll(rng))
    max_stamina = max(10, DiceFormula(2, DieType.D8, agi_mod * 2).roll(rng))
    base_armor_class = 10 + agi_mod
    speed = max(5, 20 + agi_mod * 2 + DiceFormula(1, DieType.D6).roll(rng))
    attack_bonus = str_mod
    unarmed_damage = DiceFormula(1, DieType.D6, str_mod)
    gold = DiceFormula(2, DieType.D8, 20).roll(rng)

    skills: dict[SkillType, Skill] = {}
    for type_ in SkillType.base_skills():
        primary_bonus = _CLASS_PRIMARY_BONUS if type_ in character_class.primary_skills else 0
        race_bonus = race.skill_affinities.get(type_, 0)
        level = max(1, min(MAX_SKILL_LEVEL, _BASE_SKILL_FLOOR + primary_bonus + race_bonus))
        skills[type_] = Skill(type_, level)
    if character_class.free_signature_skill is not None:
        type_ = character_class.free_signature_skill
        race_bonus = race.skill_affinities.get(type_, 0)
        level = max(1, min(MAX_SKILL_LEVEL, _SIGNATURE_SKILL_START + race_bonus))
        skills[type_] = Skill(type_, level)

    weapon = sg.weapon_item(character_class.starting_gear.weapon_name, 1, rng)
    armor = sg.armor_item(character_class.starting_gear.armor_name, 1, rng)
    # Starting gear is equipped directly (bypassing commands.py's equip(), which
    # normally bakes an item's bonus into the player's stats on equip) -- so its
    # armor_class_bonus has to be folded in here too.
    armor_class = base_armor_class + (armor.armor_class_bonus or 0)

    return PlayerCharacter(
        name=name,
        race=race,
        character_class=character_class,
        level=1,
        experience=0,
        strength=strength,
        agility=agility,
        willpower=willpower,
        max_health=max_health,
        current_health=max_health,
        max_stamina=max_stamina,
        current_stamina=max_stamina,
        armor_class=armor_class,
        speed=speed,
        attack_bonus=attack_bonus,
        unarmed_damage=unarmed_damage,
        skills=skills,
        gold=gold,
        inventory=(weapon, armor),
        equipped_weapon=weapon,
        equipped_armor=armor,
    )

File: core/src/commonMain/kotlin/eldoria/core/world/PlayerCharacterFactory.kt
package eldoria.core.world

import eldoria.core.model.CharacterClass
import eldoria.core.model.DiceFormula
import eldoria.core.model.DieType
import eldoria.core.model.PlayerCharacter
import eldoria.core.model.Race
import eldoria.core.model.Skill
import eldoria.core.model.SkillType
import eldoria.core.model.StatBlock
import kotlin.random.Random

/**
 * Builds a fresh level-1 PlayerCharacter: rolls base ability scores (3d6,
 * race-modified), derives starting combat stats with the same formulas
 * LevelProgression uses at level 1 (so leveling up is a seamless
 * continuation, not a formula switch), and starts every base (non
 * trainer-locked) skill at a level shaped by class specialization and race
 * affinity. Trainer-locked skills (magic schools, crafting) start unknown
 * except for a class's one free `freeSignatureSkill`, if it has one --
 * everything else must be learned in the world (see SkillProgression).
 */
object PlayerCharacterFactory {

    private const val BASE_SKILL_FLOOR = 15
    private const val CLASS_PRIMARY_BONUS = 20
    private const val SIGNATURE_SKILL_START = 20

    fun create(name: String, race: Race, characterClass: CharacterClass, rng: Random): PlayerCharacter {
        val strength = (DiceFormula(3, DieType.D6, 0).roll(rng) + race.abilityModifiers.strength).coerceAtLeast(1)
        val agility = (DiceFormula(3, DieType.D6, 0).roll(rng) + race.abilityModifiers.agility).coerceAtLeast(1)
        val willpower = (DiceFormula(3, DieType.D6, 0).roll(rng) + race.abilityModifiers.willpower).coerceAtLeast(1)
        val strMod = StatBlock.modifierOf(strength)
        val agiMod = StatBlock.modifierOf(agility)

        val maxHealth = DiceFormula(2, DieType.D8, strMod * 2).roll(rng).coerceAtLeast(10)
        val maxStamina = DiceFormula(2, DieType.D8, agiMod * 2).roll(rng).coerceAtLeast(10)
        val baseArmorClass = 10 + agiMod
        val speed = (20 + agiMod * 2 + DiceFormula(1, DieType.D6).roll(rng)).coerceAtLeast(5)
        val attackBonus = strMod
        val unarmedDamage = DiceFormula(1, DieType.D6, strMod)
        val gold = DiceFormula(2, DieType.D8, 20).roll(rng)

        val skills = LinkedHashMap<SkillType, Skill>()
        for (type in SkillType.baseSkills) {
            val primaryBonus = if (type in characterClass.primarySkills) CLASS_PRIMARY_BONUS else 0
            val raceBonus = race.skillAffinities[type] ?: 0
            val level = (BASE_SKILL_FLOOR + primaryBonus + raceBonus).coerceIn(1, SkillProgression.MAX_SKILL_LEVEL)
            skills[type] = Skill(type, level)
        }
        characterClass.freeSignatureSkill?.let { type ->
            val raceBonus = race.skillAffinities[type] ?: 0
            val level = (SIGNATURE_SKILL_START + raceBonus).coerceIn(1, SkillProgression.MAX_SKILL_LEVEL)
            skills[type] = Skill(type, level)
        }

        val weapon = StatGenerator.weaponItem(characterClass.startingGear.weaponName, tier = 1, rng = rng)
        val armor = StatGenerator.armorItem(characterClass.startingGear.armorName, tier = 1, rng = rng)
        // Starting gear is equipped directly (bypassing Game.kt's equip(),
        // which normally bakes an item's bonus into the player's stats on
        // equip) -- so its armorClassBonus has to be folded in here too, or
        // new characters would start with armor that mechanically does
        // nothing until the player manually re-equips it.
        val armorClass = baseArmorClass + (armor.armorClassBonus ?: 0)

        return PlayerCharacter(
            name = name,
            race = race,
            characterClass = characterClass,
            level = 1,
            experience = 0,
            strength = strength,
            agility = agility,
            willpower = willpower,
            maxHealth = maxHealth,
            currentHealth = maxHealth,
            maxStamina = maxStamina,
            currentStamina = maxStamina,
            armorClass = armorClass,
            speed = speed,
            attackBonus = attackBonus,
            unarmedDamage = unarmedDamage,
            skills = skills,
            gold = gold,
            inventory = listOf(weapon, armor),
            equippedWeapon = weapon,
            equippedArmor = armor,
        )
    }
}
